src/setup_payload/QRCodeSetupPayloadGenerator.cpp

```diff
@@ -173,8 +173,13 @@ static CHIP_ERROR generateBitSet(PayloadContents & payload, MutableByteSpan & bi
     VerifyOrReturnError(payload.rendezvousInformation.HasValue(), CHIP_ERROR_INVALID_ARGUMENT);
     ReturnErrorOnFailure(populateBits(bits.data(), offset, payload.rendezvousInformation.Value().Raw(),
                                       kRendezvousInfoFieldLengthInBits, kTotalPayloadDataSizeInBits));
-    ReturnErrorOnFailure(populateBits(bits.data(), offset, payload.discriminator.GetLongValue(),
-                                      kPayloadDiscriminatorFieldLengthInBits, kTotalPayloadDataSizeInBits));
+
+    // If the payload is valid then discriminator will always be long, but we shouldn't die if we're
+    // encoding a short one due to mAllowInvalidPayload being set. Just use the short value in that case.
+    auto const & pd        = payload.discriminator;
+    uint16_t discriminator = (pd.IsShortDiscriminator()) ? pd.GetShortValue() : pd.GetLongValue();
```

Comment:
So here's the thing: this payload is not just "invalid", it's literally nonsensical.  Does anything (including tests, etc, given that needs explicit opt-in) actually rely on encoding such payloads with short discriminators as QR codes?  If not, I don't think we should support it and should just error out.  The resulting QR code is not "invalid" in any sense, so it can't be used to test handling of invalid QR codes by other entities; it's just not a valid QR code that matches any reality in any way.

Reply:
My specific use case is to use the QR code representation to serialize MTRSetupPayload objects that may not be valid, and be able to deserialize them again. Using the QR code TLV representation seems like a much more robust way to handle the extension data than coming up with some other encoding for it.

Reply:
As discussed in Slack, I'm going to just encode the long discriminator as 0, to avoid a client making any undue assumptions about being able to roundtrip a short discriminator here.